### PR TITLE
Add partner fee meta-data

### PR DESCRIPTION
**Pull request by anxolin**
Add a new type of metadata: `partnerFee`

This meta-data will allow anyone to encode the intent to charge a fee for a given order.

- [x] Add new datata type
- [x] ~Version schema: `v0.12.0`~  Version schema: `v1.0.0` (changes are now breaking changes)
- [x] Build new types and adapt `compile.ts` to export the new metadata schema
- [x] Build tests: happy path, not-so-happy path

## Validations

- The metadata is optional, as all other metadatas
- If present, all the props are mandatory: `bip`, `recipient`
- `bip` is an integer (not a string)
- `bip` should be greater than zero, and less than 10,000 (the protocol might limit further the maximum, for example 300, or 3%)

See unit tests
image

## Example
Example on the new metadata:
```ts
partnerFee: {
   bips: 50,
   recipient: '0xb6BAd41ae76A11D10f7b0E664C5007b908bC77C9'
}
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -152,16 +152,19 @@ Fork the repo so you can create a new PR. Then:
 - Version it using the [semver](https://semver.org/) convention
 - You will need to create the new file for the meta-data schema: `<meta-data-name>/<new-version>.json`
 - Update it in the main schema you just created in step 1: Set it to `"<meta-data-name>": { "$ref": "<meta-data-name>/<new-version>.json#" }`
-- 🚨 IMPORTANT: Don't forget to add the exported constant with the latest version in:
-  - https://github.com/cowprotocol/app-data/blob/widget-metadata/src/scripts/compile.ts#L68
+
+4. Modify the `compile.ts` script
+
+- Add the exported constant with the latest version in, and the new metadata:
+  - For example: https://github.com/cowprotocol/app-data/pull/44/commits/aeef8a58e7bbd2a53664ce396011cb157a18406d
```
Comment (anxolin) [inline README.md:159]: unrelated: this link was just broken

**Review by alfetopito — CHANGES_REQUESTED** [review]
Other than the comments, the build is currently broken

**Comment by anxolin** [comment]
The build was broken before. Fixing in https://github.com/cowprotocol/app-data/pull/54

File: src/schemas/definitions.json [hunk 2]
```diff
@@ -7,32 +7,32 @@
...
-      "examples": [
-        "90741097240912730913",
-        "0",
-        "75891372"
-      ],
+      "examples": ["90741097240912730913", "0", "75891372"],
       "type": "string"
+    },
+    "bips": {
+      "$id": "#/definitions/bips",
```
[the remark this answers is not in the document]
Comment (anxolin) [reply src/schemas/definitions.json:29]: I did the other way around here: https://github.com/cowprotocol/app-data/pull/55

File: src/scripts/compile.ts [hunk 3]
```diff
@@ -76,6 +76,7 @@ export const LATEST_UTM_METADATA_VERSION = '${extractSemver(latestUtmVersion)}'
 export const LATEST_HOOKS_METADATA_VERSION = '${extractSemver(latestHooksVersion)}'
 export const LATEST_SIGNER_METADATA_VERSION = '${extractSemver(latestSignerVersion)}'
 export const LATEST_WIDGET_METADATA_VERSION = '${extractSemver(latestWidgetVersion)}'
+export const LATEST_PARTNER_FEE_METADATA_VERSION = '${extractSemver(latestWidgetVersion)}'
```
[the remark this answers is not in the document]
Comment (anxolin) [reply src/scripts/compile.ts:79]: Fixed 0d13cb8f7ad8614e982034371b5f70b8328df72a

File: test/schema.spec.ts [hunk 4]
```diff
@@ -627,6 +628,142 @@ describe('Schema v0.11.0', () => {
   )
 })
 
+describe('Schema v0.12.0', () => {
```
Comment (alfetopito) [inline test/schema.spec.ts:631]: As mentioned in #55, v0.12 won't be created. It'll be v1.0 instead

**Comment by fleupold** [comment]
Can we call it bps. Bips is very colloquial and not the official abbreviation: https://www.investopedia.com/ask/answers/what-basis-point-bps

**Comment by anxolin** [comment]
@fleupold no problem with that on my end. The only issue will be that for coherence would be great if I change it also in the quote information. I could make it part of the breaking change https://github.com/cowprotocol/app-data/pull/55 

This would require @olgafetisova to adapt the queries, and I'm unsure if that's problematic or not. Maybe she could confirm if it's ok to rename.

**Comment by alfetopito** [comment]
> @fleupold no problem with that on my end. The only issue will be that for coherence would be great if I change it also in the quote information. I could make it part of the breaking change #55
> 
> This would require @olgafetisova to adapt the queries, and I'm unsure if that's problematic or not. Maybe she could confirm if it's ok to rename.

Keep in mind this change would require analytics to handle both cases for `quotes`

**Comment by anxolin** [comment]
> Keep in mind this change would require analytics to handle both cases for quotes

yes, that was my point. This is why I bring it up:

- Option 1: keep all as defined in #53 and #54 (coherent, using the colloquial term)
- Option 2: use bps in `parnerFee` but not in `quote`  (incoherent, using the right term at least in the partner fee)
- Option 3: use bps in both `parnerFee` and quote `quote` (coherent, using the right term, potentially adding trouble for analytics to adapt queries or discard older data in some queries)

File: test/schema.spec.ts (hunk 4 above)
Comment (anxolin) [reply test/schema.spec.ts:631]: Okay, but the changes are independent and I did the version here. I thought is ok to version it still, as those are 2 different changes and the one in this PR is backward compatible

I will delete the 0.12.0 in #54 instead as it seems you prefer 0.12.0 don't exist. If fine with that

Comment (anxolin) [reply test/schema.spec.ts:631]: a4789f02b2c2e7a852ee52febacb1e2368c45428

**Comment by olgafetisova** [comment]
> This would require @olgafetisova to adapt the queries, and I'm unsure if that's problematic or not. Maybe she could confirm if it's ok to rename.

We are currently changing slippageBips to integer in the [code](https://github.com/duneanalytics/spellbook/blob/788bf984c292f9311dc0394ff6f97dfe5f6f8cec/models/cow_protocol/ethereum/cow_protocol_ethereum_app_data.sql#L43) right away before doing any other adjustments, so I think casting both string and integer to integer should not break anything.

**Comment by alfetopito** [comment]
> > Keep in mind this change would require analytics to handle both cases for quotes
> 
> yes, that was my point. This is why I bring it up:
> 
>     * Option 1: keep all as defined in [Add partner fee meta-data #53](https://github.com/cowprotocol/app-data/pull/53) and [Fix build #54](https://github.com/cowprotocol/app-data/pull/54) (coherent, using the colloquial term)
> 
>     * Option 2: use bps in `parnerFee` but not in `quote`  (incoherent, using the right term at least in the partner fee)
> 
>     * Option 3: use bps in both `parnerFee` and quote `quote` (coherent, using the right term, potentially adding trouble for analytics to adapt queries or discard older data in some queries)

Due to historical reasons and consistency, I prefer `option 1`.
Unless @olgafetisova doesn't mind the additional change/handling to go for `option 3`

**Comment by olgafetisova** [comment]
> > ```
> > * Option 1: keep all as defined in [Add partner fee meta-data #53](https://github.com/cowprotocol/app-data/pull/53) and [Fix build #54](https://github.com/cowprotocol/app-data/pull/54) (coherent, using the colloquial term)
> > 
> > * Option 2: use bps in `parnerFee` but not in `quote`  (incoherent, using the right term at least in the partner fee)
> > 
> > * Option 3: use bps in both `parnerFee` and quote `quote` (coherent, using the right term, potentially adding trouble for analytics to adapt queries or discard older data in some queries)
> > ```

If you change slippageBips to bips then it will break some queries, not a big work to adjust it but might take 1-3 days from Dune side to approve the PR, so we might be experiencing some issues there. So I would prefer the name to stay as it is for slippage. 

If you change slippageBips from string to integer, it won't break the code. Everything will remain as is.

**Comment by anxolin** [comment]
I would personally not add you 1-3 days of work for this name change (discard option 3) @fleupold do you agree? 

If we can't change it for the quote, I would slightly prefer option 2, yes it's not coherent but at least we don't keep using the wrong term for some metadata that is more relevant for integrators. 

Otherwise, we can go with option 1.

**Comment by olgafetisova** [comment]
I went over the documentation again and  I do not see any simple way to allow a non-existing column in the code (as it will happen when we add the change). The change has to happen on the ETL side which is written in a flat way (same as in our backend), hence, there is no consideration for columns.
To conclude, my strong preference would be to keep the name as is as I did not find an easy way to adjust the code.

**Comment by anxolin** [comment]
Okay, I believe we arrive to consensus over option 2. I applied the changes. Will create a release of this library

**Review by alfetopito — COMMENTED** [review]
👍

**Pull request merged**